### PR TITLE
oauth works when creating users now

### PR DESCRIPTION
**Pull request by nadrane**

### CODE REVIEW

**Comment by nadrane** [comment]
@glebec can you check this real fast?

**Comment by glebec** [comment]
@nadrane I think @intersim already fixed this in #59. You can see the current line at https://github.com/FullstackAcademy/bones/blob/master/db/models/oauth.js#L60.

**Comment by nadrane** [comment]
Oh, is it just not in master?

**Comment by nadrane** [comment]
I could've sworn I pulled

**Comment by glebec** [comment]
I'm looking at it in master (the link above is to master). Odd.

**Comment by glebec** [comment]
Oh wait. You're _adding_ the deconstruction back in.

**Comment by nadrane** [comment]
haha yes, it doesn't work without it. Serialization crashes because we try to grab the id of something that isn't a sequelize user instance

**Comment by nadrane** [comment]
Instead, what we pass to the serialization function is an object containing a sequelize user and an ouath instance

**Comment by glebec** [comment]
That doesn't seem to make sense to me. The previous `then` returns a user instance.

I'll pull the code down and play with it to check.

**Comment by nadrane** [comment]
The previous .then returns the result of this promise:

```
db.Promise.props({
        user,
        _setOauthUser: oauth.setUser(user)
      })
```

As I understand it, that promise resolves to an object for 2 sequelize instances (one user, one ouath), not just a user instance

**Comment by glebec** [comment]
That's not what this PR is showing. The previous `then` is this:

```js
.then(({ oauth, user }) => user ||
  User.create({
    name: profile.displayName,
  }).then(user => db.Promise.props({
    user,
    _setOauthUser: oauth.setUser(user)
  }))
)
```

**Comment by glebec** [comment]
Oh I see, there's a nested promise chain. Blah

**Comment by glebec** [comment]
So wait, that means this PR is still not enough. Because sometimes a `user` is returned, and sometimes a `props` promise is returned. So this is really an issue of the promise chain sometimes returning different things.

**Comment by glebec** [comment]
An ugly way to fix could be something like:

```js
.then(({ oauth, user }) => user ||
  User.create({
    name: profile.displayName,
  }).then(user => db.Promise.props({
    user,
    _setOauthUser: oauth.setUser(user)
  })).then(({user}) => user))
)
```

**Comment by nadrane** [comment]
oh, good point

**Comment by glebec** [comment]
It seems to me that at least some of this promise chain is just trying to pass already-resolved data down the chain. That can be handled more directly with Bluebird's `bind` method so that data needed downstream can be set on `this`.

**Comment by nadrane** [comment]
can we just teach the students async/await instead :)

**Comment by glebec** [comment]
Do you want to complete the fix? I have a couple other things I need to do before I can sleep, and I haven't set up an auth testing flow yet.

**Comment by nadrane** [comment]
I don't think I'm going to fix this tonight. i want to make sure whatever I submit works and leave time to test it properly. We should just make sure everyone knows there's an issue

**Comment by glebec** [comment]
I just pushed a fix which I tested with Google OAuth. I don't want to merge it though as I'm tired and want someone to run a sanity check on it. I'll leave some line comments to highlight specifics, because I also modified the formatting slightly to be easier to read.

File: db/models/oauth.js [hunk 1]
```diff
@@ -32,33 +32,38 @@ OAuth.V2 = (accessToken, refreshToken, profile, done) =>
...
-      oauth.profileJson = profile
-      oauth.accessToken = accessToken
+    }
+  })
+  .spread(oauth => {
+    console.log(profile)
+    debug('provider:%s will log in user:{name=%s uid=%s}',
+      profile.provider,
+      profile.displayName,
+      profile.id
```
Comment (glebec) [inline db/models/oauth.js]: `profile` doesn't have a `uid` — `oauth` does. Fixed to `profile.id`.

File: db/models/oauth.js [hunk 2]
```diff
@@ -32,33 +32,38 @@ OAuth.V2 = (accessToken, refreshToken, profile, done) =>
...
+  })
+  .then(({ oauth, user }) => user ||
+    User.create({
+      name: profile.displayName,
+    })
+    .then(user => db.Promise.props({
+      user,
+      _setOauthUser: oauth.setUser(user)
+    }))
+    .then(({user}) => user)
```
Comment (glebec) [inline db/models/oauth.js]: this line, with line 55, means that this step in the parent promise chain always returns a usable `user` object.

File: server/auth.js [hunk 3]
```diff
@@ -127,6 +127,7 @@ auth.post('/login/local', passport.authenticate('local', { successRedirect: '/',
 // Register this route as a callback URL with OAuth provider
 auth.get('/login/:strategy', (req, res, next) =>
   passport.authenticate(req.params.strategy, {
+    scope: 'email',
```
Comment (glebec) [inline server/auth.js]: Google OAuth requires an `email` scope at minimum.

*I have not tested this with other providers.* Please double-check that it makes sense to add this scope even for other providers.

Comment (intersim) [inline server/auth.js]: Still getting a weird error for github, but this works for Google... gonna merge, but please be advised!

**Pull request merged**